pwnlib/heap/glmalloc/heap_explorer.py

```diff
@@ -0,0 +1,698 @@
+        # by safe-link
+        self.safe_link = safe_link
+
+    def _is_safe_link_enabled(self):
+        # safe-link protection is included in version 2.32
+        return self._process_informer.is_libc_version_higher_than((2, 31))
+
+    def _are_tcaches_enabled(self):
+        # tcaches were added in version 2.26
+        return self._process_informer.is_libc_version_higher_than((2, 25))
+
+    def arenas_count(self):
+        """Returns the number of arenas
+
+        Returns:
+            int
+
+        Examples:
+            >>> p = process('sh')
+            >>> hp = p.heap_explorer()
+            >>> number_of_arenas = hp.arenas_count()
+            >>> number_of_arenas # doctest: +SKIP
+            2
+        """
+        return len(self.all_arenas_malloc_states())
+
+    def malloc_state(self, arena_index=0):
+        """Returns the malloc_arena of the arena
+
+        Args:
+            arena_index (int, optional): The index of the desired arena. If none
+                is specified, then the index of the main arena will be selected
+
+        Returns:
+            :class:`MallocState`
+
+        Examples:
+            >>> p = process('sh')
+            >>> hp = p.heap_explorer()
+            >>> ms = hp.malloc_state()
+            >>> ms_top = ms.top
+            >>> hex(ms_top) # doctest: +ELLIPSIS
+            '0x...'
+            >>> ms_address = ms.address
+            >>> hex(ms_address) # doctest: +ELLIPSIS
+            '0x...'
+            >>> ms_str = str(ms)
+            >>> print(ms_str) # doctest: +ELLIPSIS
+            ====... Malloc State (0x...) ====...
+            mutex = 0x...
+            flags = 0x...
+            have_fastchunks = 0x...
+            fastbinsY
+              [0] 0x20 => 0x...
+              [1] 0x30 => 0x...
+              [2] 0x40 => 0x...
+              [3] 0x50 => 0x...
+              [4] 0x60 => 0x...
+              [5] 0x70 => 0x...
+              [6] 0x80 => 0x...
+              [7] 0x90 => 0x...
+              [8] 0xa0 => 0x...
+              [9] 0xb0 => 0x...
+            top = 0x...
+            last_remainder = 0x...
+            bins
+             Unsorted bins
+              [0] fd=0x... bk=0x...
+             Small bins
+              [1] 0x20 fd=0x... bk=0x...
+              [2] 0x30 fd=0x... bk=0x...
+            ...
+             Large bins
+              [63] 0x400 fd=0x... bk=0x...
+              [64] 0x440 fd=0x... bk=0x...
+            ...
+            binmap = [0x..., 0x..., 0x..., 0x...]
+            next = 0x...
+            next_free = 0x...
+            attached_threads = 0x...
+            system_mem = 0x...
+            max_system_mem = 0x...
+            ==========================...
+        """
+        return self.all_arenas_malloc_states()[arena_index]
+
+    def all_arenas_malloc_states(self):
+        """Returns the malloc states of all arenas
+
+        Returns:
+            list of :class:`MallocState`
+
+        Examples:
+            >>> p = process('sh')
```

Comment:
I would much rather see these tests operate on a Corefile.  Create a new module in `pwnlib/data` (see the existing ones for an example) like `pwnlib/data/heap/foo.core` and then you can:

```
Corefile(pwnlib.data.heap.get('foo.core'))
```

And you won't need the ellipsis anymore.

Reply:
Done in #ea696ca